gensim/models/keyedvectors.py

```diff
@@ -220,7 +220,7 @@ def add_word(word, weights):
                 result.index2word.append(word)
 
             if binary:
-                binary_len = dtype(REAL).itemsize * vector_size
+                binary_len = dtype(datatype).itemsize * vector_size
```

Comment:
I am having problem detecting `binary_len` of vector saved with custom datatype. The only clue is that the next vector starts after a `" "` but before the space comes a string(also converted in python bytes) which can be of any length. @menshikh-iv any suggestion?

Reply:
I tried changing `\n` at the end of each vector after each vector but that broke many other tests.

Reply:
@pushpankar how it works if `datatype=REAL` here?

Reply:
Since earlier float were only saved with only 32 bit precision, knowing the size of each vector in binary format was easy. Casting to lower precision is only done after loading vectors.
Please also note that in develop branch too, casting vector to lower precision and saving it in binary and then loading leads to some errors. This is because while loading float32 is being assumed but during saving it was saved with lower precision like float16. I am adding some code to make it more clear.
```
from gensim.models.keyedvectors import KeyedVectors
model = KeyedVectors.load_word2vec_format('./test_data/test.kv.txt', datatype=np.float16)
print(model['horse.n.01'][0])
model.save_word2vec_format('./test_data/test.kv.bin', binary=True)
model2 = KeyedVectors.load_word2vec_format('./test_data/test.kv.bin', datatype=np.float32, binary=True)
print(model2['horse.n.01'][0])
```
Gives 
```
Traceback (most recent call last):
  File "convert2binary.py", line 7, in <module>
    print(model2['horse.n.01'][0])
  File "/home/pushpankar/gensim/gensim/models/keyedvectors.py", line 326, in __getitem__
    return self.word_vec(words)
  File "/home/pushpankar/gensim/gensim/models/keyedvectors.py", line 453, in word_vec
    raise KeyError("word '%s' not in vocabulary" % word)
KeyError: "word 'horse.n.01' not in vocabulary"

```
This is because float32 was assumed while reading binary vector but originally it was saved with float16. Thus more than necessary bytes was read for every vector.
Let me know if I am not clear enough.

Reply:
so, probably the easiest solution for this case is read/write with `REAL` type & cast it before the end of "load" process, wdyt @jayantj?